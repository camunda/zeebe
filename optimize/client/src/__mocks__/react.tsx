/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

// @ts-ignore
export * from 'react';
export {default} from 'react';

const outstandingEffects: (() => void)[] = [];
const outstandingCleanups: (() => void)[] = [];

export const useEffect = (fn: () => (() => void) | void) => outstandingEffects.push(fn);
export const runLastEffect = () => {
  if (outstandingEffects.length) {
    const cleanup = outstandingEffects.pop()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

export const runLastCleanup = () => {
  if (outstandingCleanups.length) {
    outstandingCleanups.pop()?.();
  }
};

export const runAllEffects = () => {
  const numberOfEffects = outstandingEffects.length;

  for (let i = 0; i < numberOfEffects; i++) {
    const cleanup = outstandingEffects.shift()?.();

    if (cleanup) {
      outstandingCleanups.push(cleanup);
    }
  }
};

beforeEach(() => {
  outstandingEffects.length = 0;
});

export const runAllCleanups = () => {
  const numberOfEffects = outstandingCleanups.length;

  for (let i = 0; i < numberOfEffects; i++) {
    outstandingCleanups.shift()?.();
  }
};
