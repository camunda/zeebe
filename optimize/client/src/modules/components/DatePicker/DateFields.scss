.DateFields {
  position: relative;

  .inputContainer {
    display: flex;
    align-items: flex-start;
    flex-wrap: nowrap;

    .PickerDateInput:first-child {
      margin-right: 15px;
    }

    .PickerDateInput:nth-child(2):before {
      left: -10px;
      position: absolute;
      top: 50%;
      transform: translateY(-50%);
      content: '-';
    }
  }

  .dateRangeContainer {
    position: absolute;
    width: max-content;
    left: 0;
    top: 50px;
    box-shadow: 0 5px 15px rgba(0, 0, 0, 0.5);
    border: 1px solid rgba(0, 0, 0, 0.2);
    background-color: var(--white);
    padding: 5px 0;
    z-index: var(--z-popup);

    &.dateRangeContainerLeft {
      &::before {
        left: 50px;
      }

      &::after {
        left: 30px;
      }
    }

    &.dateRangeContainerRight {
      &::before {
        left: 230px;
      }
      &::after {
        left: 210px;
      }
    }

    &::before {
      content: ' ';
      display: block;
      width: 20px;
      height: 20px;
      position: absolute;
      transform: rotate(45deg);
      border: 1px solid rgba(0, 0, 0, 0.2);
      background-color: var(--white);
      top: -10px;
    }

    &::after {
      content: ' ';
      display: block;
      width: 50px;
      height: 22px;
      position: absolute;
      background-color: var(--white);
      top: 0;
    }
  }
}
