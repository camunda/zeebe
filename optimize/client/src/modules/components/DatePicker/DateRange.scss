.dark {
  .rdrNextPrevButton {
    background: var(--grey-darken-36);
  }

  .rdrNextPrevButton:hover {
    background: var(--grey-darken-23);
  }

  .rdrPprevButton i {
    border-color: transparent var(--silver-darken-87) transparent transparent;
  }

  .rdrNextButton i {
    border-color: transparent transparent transparent var(--silver-darken-87);
  }

  .rdrMonthAndYearPickers select {
    color: var(--silver-darken-87);
    background: initial;
    appearance: menulist;

    option {
      background-color: var(--grey-darken-23);
    }
  }

  .rdrMonthAndYearPickers select:hover {
    background-color: rgba(0, 0, 0, 0.07);
  }

  .rdrMonthName,
  .rdrWeekDay {
    color: var(--grey-lighten-56);
  }

  .rdrDayPassive .rdrDayNumber span {
    color: var(--grey-lighten-56);
  }

  .rdrDayNumber span {
    color: var(--silver-darken-87);
  }
}

.rdrCalendarWrapper .rdrDateDisplayWrapper {
  display: none;
}
