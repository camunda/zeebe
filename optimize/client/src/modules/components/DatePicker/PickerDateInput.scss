.PickerDateInput {
  position: relative;

  .cds--text-input {
    width: 150px;
    padding-right: 2rem;
  }

  .icon {
    position: absolute;
    top: 50%;
    transform: translateY(-50%);
    right: 1rem;
    pointer-events: none;
  }

  .cds--form-requirement {
    position: absolute;
    top: 40px;
  }

  &.isInvalid {
    .icon {
      display: none;
    }
  }
}
