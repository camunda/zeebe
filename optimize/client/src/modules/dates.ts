/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {format as formatFns, Locale} from 'date-fns';

export let globalLocale: Locale;

// by providing a default string of 'PP' or any of its variants for `formatStr`
// it will format dates in whichever way is appropriate to the locale
export function format(date: Date | number, formatStr = 'PP'): string {
  return formatFns(date, formatStr, {
    locale: globalLocale,
  });
}

export async function loadDateTranslation(localeCode: string): Promise<void> {
  switch (localeCode) {
    case 'de':
      globalLocale = (await import('date-fns/locale/de')).default;
      break;
    case 'fr':
      globalLocale = (await import('date-fns/locale/fr')).default;
      break;
    case 'es':
      globalLocale = (await import('date-fns/locale/es')).default;
      break;
    case 'hi':
      globalLocale = (await import('date-fns/locale/hi')).default;
      break;
    case 'it':
      globalLocale = (await import('date-fns/locale/it')).default;
      break;
    case 'ja':
      globalLocale = (await import('date-fns/locale/ja')).default;
      break;
    case 'pt':
      globalLocale = (await import('date-fns/locale/pt')).default;
      break;
    case 'ru':
      globalLocale = (await import('date-fns/locale/ru')).default;
      break;
    default:
      break;
  }
}

export const BACKEND_DATE_FORMAT = "yyyy-MM-dd'T'HH:mm:ss.SSSxx";
