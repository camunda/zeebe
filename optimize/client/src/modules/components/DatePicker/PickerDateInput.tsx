/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {ComponentProps, ForwardedRef, forwardRef, memo} from 'react';
import {TextInput} from '@carbon/react';
import {Calendar} from '@carbon/icons-react';
import classnames from 'classnames';

import {t} from 'translation';

import './PickerDateInput.scss';

interface PickerDateInputProps extends Omit<ComponentProps<typeof TextInput>, 'onChange'> {
  reference?: ForwardedRef<HTMLInputElement>;
  onChange: (value: string) => void;
  onSubmit: () => void;
}

export const PickerDateInput = memo(
  ({onChange, onSubmit, reference, invalid, ...props}: PickerDateInputProps) => {
    const isInvalid = !!props.value && invalid;
    return (
      <div className={classnames('PickerDateInput', {isInvalid})}>
        <TextInput
          {...props}
          invalid={isInvalid}
          invalidText={t('common.filter.dateModal.invalidDate')}
          placeholder="yyyy-mm-dd"
          onChange={({target: {value}}) => onChange(value)}
          onKeyDown={({key}) => key === 'Enter' && onSubmit()}
          ref={reference}
        />
        <Calendar className="icon" />
      </div>
    );
  }
);

export default forwardRef<HTMLInputElement, PickerDateInputProps>((props, ref) => (
  <PickerDateInput {...props} reference={ref} />
));
