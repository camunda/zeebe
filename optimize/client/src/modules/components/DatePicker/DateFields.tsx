/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {KeyboardEventHandler, PureComponent, createRef} from 'react';
import classnames from 'classnames';
import {parseISO} from 'date-fns';

import {format} from 'dates';
import {t} from 'translation';

import DateRange from './DateRange';
import PickerDateInput from './PickerDateInput';
import {isDateValid} from './service';

import './DateFields.scss';

const POPUP_CLASSNAME = 'dateRangeContainer';

interface DateFieldsProps {
  startDate: string;
  endDate: string;
  forceOpen?: boolean;
  type: 'between' | 'after' | 'before';
  format: string;
  onDateChange: (type: DateFieldName, value: string) => void;
}

export type DateFieldName = 'startDate' | 'endDate';

interface DateFieldsState {
  popupOpen: boolean;
  currentlySelectedField: DateFieldName | null;
}

export default class DateFields extends PureComponent<DateFieldsProps, DateFieldsState> {
  state: DateFieldsState = {
    popupOpen: false,
    currentlySelectedField: null,
  };

  dateFields = createRef<HTMLDivElement>();

  endDateField = createRef<HTMLInputElement>();

  // Modals stop propagation of events on elements outside the modal
  // Therefore, we need to attach the events on the modal instead of document
  getContext = () => this.dateFields.current?.closest<HTMLElement>('.Modal') || document;

  componentDidUpdate() {
    const {popupOpen} = this.state;

    const context = this.getContext() as HTMLElement;

    if (popupOpen) {
      context.addEventListener('click', this.hidePopup);
      context.addEventListener('keydown', this.closeOnEscape);
    } else {
      context.removeEventListener('click', this.hidePopup);
      context.removeEventListener('keydown', this.closeOnEscape);
    }
  }

  componentWillUnmount() {
    const context = this.getContext() as HTMLElement;
    context.removeEventListener('click', this.hidePopup);
    context.removeEventListener('keydown', this.closeOnEscape);
  }

  handleKeyPress: KeyboardEventHandler = (evt) => {
    if (this.state.popupOpen && evt.key === 'Escape') {
      evt.stopPropagation();
    }
  };

  render() {
    const {startDate, endDate, forceOpen, type} = this.props;

    const startDateObj = parseISO(startDate);
    const endDateObj = parseISO(endDate);

    return (
      <div className="DateFields" ref={this.dateFields} onKeyDown={this.handleKeyPress}>
        <div className="inputContainer">
          {type !== 'before' && (
            <PickerDateInput
              id="dateFieldsBefore"
              labelText={t('common.filter.dateModal.unit.before')}
              hideLabel
              onChange={this.setDate('startDate')}
              onFocus={() => {
                this.setState({currentlySelectedField: 'startDate'});
              }}
              onSubmit={this.submitStart}
              onClick={() => this.toggleDateRangePopup('startDate')}
              value={startDate}
              invalid={!isDateValid(startDate)}
            />
          )}
          {type !== 'after' && (
            <PickerDateInput
              id="dateFieldsAfter"
              labelText={t('common.filter.dateModal.unit.after')}
              hideLabel
              ref={this.endDateField}
              onChange={this.setDate('endDate')}
              onFocus={() => {
                this.setState({currentlySelectedField: 'endDate'});
              }}
              onSubmit={this.submitEnd}
              onClick={() => this.toggleDateRangePopup('endDate')}
              value={endDate}
              invalid={!isDateValid(endDate)}
            />
          )}
        </div>
        {(this.state.popupOpen || forceOpen) && (
          <div
            className={classnames(POPUP_CLASSNAME, {
              dateRangeContainerLeft:
                this.isFieldSelected('startDate') ||
                (this.isFieldSelected('endDate') && type === 'before'),
              dateRangeContainerRight: this.isFieldSelected('endDate') && type !== 'before',
            })}
          >
            <DateRange
              type={type}
              endDateSelected={this.isFieldSelected('endDate')}
              onDateChange={this.onDateRangeChange}
              startDate={startDateObj}
              endDate={endDateObj}
            />
          </div>
        )}
      </div>
    );
  }

  submitStart = () => {
    if (this.props.type === 'between') {
      this.setState({currentlySelectedField: 'endDate'});
      this.endDateField.current?.focus();
    } else {
      this.hidePopup();
    }
  };

  submitEnd = () => {
    this.hidePopup();
    this.endDateField.current?.blur();
  };

  closeOnEscape = (event: KeyboardEvent) => {
    if (event.key === 'Escape') {
      this.hidePopup();
    }
  };

  formatDate = (date?: Date | null) => (date ? format(date, this.props.format) : '');

  onDateRangeChange = ({
    startDate,
    endDate,
  }: {startDate?: Date | null; endDate?: Date | null} = {}) => {
    this.props.onDateChange('startDate', this.formatDate(startDate));
    this.props.onDateChange('endDate', this.formatDate(endDate));

    if (this.isFieldSelected('endDate') || this.props.type !== 'between') {
      setTimeout(this.hidePopup, 350);
    } else {
      this.setState({currentlySelectedField: 'endDate'});
      this.endDateField.current?.focus();
    }
  };

  hidePopup = (evt?: Event) => {
    if (!(evt?.target as HTMLElement)?.closest('.' + POPUP_CLASSNAME)) {
      this.setState({
        popupOpen: false,
        currentlySelectedField: null,
      });
    }
  };

  isFieldSelected(field: DateFieldName): boolean {
    return this.state.currentlySelectedField === field;
  }

  setDate = (name: DateFieldName) => (date: string) => this.props.onDateChange(name, date);

  toggleDateRangePopup = (field: DateFieldName) => {
    this.setState({
      popupOpen: true,
      currentlySelectedField: field,
    });
  };
}
