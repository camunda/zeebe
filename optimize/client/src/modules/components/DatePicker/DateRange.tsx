/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {Calendar, DateRange as ReactDateRange} from 'react-date-range';
import {isValid, isAfter, isEqual} from 'date-fns';

import {getLanguage} from 'translation';
import {globalLocale} from 'dates';

import 'react-date-range/dist/styles.css'; // main style file
import 'react-date-range/dist/theme/default.css'; // theme css file
import './DateRange.scss';

interface DateRangeProps {
  type: 'between' | 'after' | 'before';
  startDate?: Date | null;
  endDate?: Date | null;
  onDateChange: (range?: {startDate?: Date | null; endDate?: Date | null}) => void;
  endDateSelected?: boolean;
}

export default function DateRange({
  type,
  startDate,
  endDate,
  onDateChange,
  endDateSelected,
}: DateRangeProps) {
  const validStartDate = startDate && isValid(startDate) ? startDate : new Date();
  const validEndDate = endDate && isValid(endDate) ? endDate : new Date();

  let range;
  if (
    startDate &&
    endDate &&
    isValid(startDate) &&
    isValid(endDate) &&
    (isEqual(startDate, endDate) || isAfter(endDate, startDate))
  ) {
    range = {
      startDate,
      endDate,
    };
  } else {
    range = {
      startDate: new Date(),
      endDate: new Date(),
    };
  }

  return type === 'between' ? (
    <ReactDateRange
      focusedRange={endDateSelected ? [0, 1] : [0, 0]}
      locale={getLanguage() !== 'en' ? globalLocale : undefined}
      ranges={[range]}
      months={2}
      onChange={({range1}) => onDateChange(range1)}
      direction="horizontal"
      shownDate={endDateSelected ? range.endDate : range.startDate}
    />
  ) : (
    <Calendar
      date={type === 'after' ? validStartDate : validEndDate}
      locale={getLanguage() !== 'en' ? globalLocale : undefined}
      onChange={(date) => {
        if (type === 'after') {
          onDateChange({startDate: date, endDate: null});
        } else {
          onDateChange({startDate: null, endDate: date});
        }
      }}
    />
  );
}
