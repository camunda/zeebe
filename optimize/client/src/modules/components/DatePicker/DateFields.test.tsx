/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. Licensed under a proprietary license.
 * See the License.txt file for more information. You may not use this file
 * except in compliance with the proprietary license.
 */

import {RefObject} from 'react';
import {shallow} from 'enzyme';
import {parseISO} from 'date-fns';

import {format} from 'dates';

import DateFields from './DateFields';
import DateRange from './DateRange';
import PickerDateInput from './PickerDateInput';

jest.useFakeTimers();

const dateFormat = 'yyyy-MM-dd';

const props = {
  startDate: format(parseISO('2017-08-29'), dateFormat),
  endDate: format(parseISO('2020-06-05'), dateFormat),
  format: dateFormat,
  type: 'between' as const,
  onDateChange: jest.fn(),
};

it('should match snapshot', () => {
  const node = shallow(<DateFields {...props} />);

  expect(node).toMatchSnapshot();
});

it('should set startDate on date change of start date input field', () => {
  const spy = jest.fn();
  const node = shallow<DateFields>(<DateFields {...props} onDateChange={spy} />);

  node.instance().setDate('startDate')('change');

  expect(spy).toBeCalledWith('startDate', 'change');
});

it('should set endDate on date change of end date input field', () => {
  const spy = jest.fn();
  const node = shallow<DateFields>(<DateFields {...props} onDateChange={spy} />);

  node.instance().setDate('endDate')('change');

  expect(spy).toBeCalledWith('endDate', 'change');
});

it('should select date range popup on date input click', () => {
  const node = shallow(<DateFields {...props} />);

  node.find(PickerDateInput).at(0).simulate('click');

  expect(node.state('popupOpen')).toBe(true);
  expect(node.state('currentlySelectedField')).toBe('startDate');
});

it('should have DateRange', () => {
  const node = shallow(<DateFields {...props} />);
  node.setState({popupOpen: true});

  expect(node.find(DateRange)).toExist();
});

it('should update start and end date when selecting a date', () => {
  const spy = jest.fn();
  const node = shallow<DateFields>(<DateFields {...props} onDateChange={spy} />);

  node.setState({popupOpen: true, currentlySelectedField: 'startDate'});

  node.instance().endDateField = {
    current: {focus: jest.fn()},
  } as unknown as RefObject<HTMLInputElement>;
  node.instance().onDateRangeChange({startDate: new Date(), endDate: new Date()});
  const today = format(new Date(), dateFormat);
  expect(spy).toHaveBeenCalledWith('startDate', today);
  expect(spy).toHaveBeenCalledWith('endDate', today);
  expect(node.state('currentlySelectedField')).toBe('endDate');
});

it('should not close popup when clicking inside it', () => {
  const node = shallow<DateFields>(<DateFields {...props} />);

  node.setState({popupOpen: true, currentlySelectedField: 'startDate'});

  node.instance().hidePopup({target: {closest: () => true}} as unknown as Event);

  expect(node.state('popupOpen')).toBe(true);
});

it('should close popover after selecting date if type is not "between"', () => {
  const node = shallow(<DateFields {...props} onDateChange={jest.fn()} type="after" />);

  node.setState({popupOpen: true});

  node.find(DateRange).prop('onDateChange')({startDate: new Date(), endDate: new Date()});

  jest.runAllTimers();

  expect(node.state('popupOpen')).toBe(false);
});
