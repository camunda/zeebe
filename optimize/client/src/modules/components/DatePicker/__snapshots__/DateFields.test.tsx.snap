// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<div
  className="DateFields"
  onKeyDown={[Function]}
>
  <div
    className="inputContainer"
  >
    <ForwardRef
      hideLabel={true}
      id="dateFieldsBefore"
      invalid={false}
      labelText="Before"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
      value="2017-08-29"
    />
    <ForwardRef
      hideLabel={true}
      id="dateFieldsAfter"
      invalid={false}
      labelText="After"
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onSubmit={[Function]}
      value="2020-06-05"
    />
  </div>
</div>
`;
